feat: add pipeline registry abstraction

aarnphm opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

- added `PipelineRegistry` abstraction for better supports for custom pipeline. 
- updates `add_new_pipeline.mdx` (english docs) to reflect the api addition
- migrate `check_task` and `get_supported_tasks` from
  `transformers/pipelines/__init__.py` to
 ` transformers/pipelines/base.py#PipelineRegistry.{check_task,get_supported_tasks}`

Address #17762


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@LysandreJik and @sgugger, would be great if you guys can provide feedback.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

aarnphm committed:
feat: add pipeline registry abstraction

- added `PipelineRegistry` abstraction
- updates `add_new_pipeline.mdx` (english docs) to reflect the api addition
- migrate `check_task` and `get_supported_tasks` from
  transformers/pipelines/__init__.py to
  transformers/pipelines/base.py#PipelineRegistry.{check_task,get_supported_tasks}

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm committed:
fix: update with upstream/main

chore: Apply suggestions from sgugger's code review

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger commented on src/transformers/pipelines/__init__.py:
```diff
@@ -305,14 +305,14 @@
     elif values["type"] != "multimodal":
         raise ValueError(f"SUPPORTED_TASK {task} contains invalid type {values['type']}")
 
+PipelineRegistry = _PipelineRegistry.from_dict(supported_tasks=SUPPORTED_TASKS, task_aliases=TASK_ALIASES)  # type: ignore (possible unbound type due to lazy module import)
```
```suggestion
PipelineRegistry = _PipelineRegistry.from_dict(supported_tasks=SUPPORTED_TASKS, task_aliases=TASK_ALIASES)
```

sgugger reviewed (commented):
Thanks a lot for your PR! This could be an interesting addition. I'm curious to see what @Narsil and @LysandreJik think of it.

Note that this is only to customize preprocessing and/or postprocessing as this still relies on existing auto-model classes.

aarnphm commented on src/transformers/pipelines/base.py:
```diff
@@ -1085,3 +1085,48 @@ def get_iterator(
+            task = self.TASK_ALIASES[task]
+        if task in self.SUPPORTED_TASKS:
+            targeted_task = self.SUPPORTED_TASKS[task]
+            return targeted_task, None  # type: ignore (unfinshed types)
```
sounds good. I will remove accordingly.

aarnphm commented:
> Note that this is only to customize preprocessing and/or postprocessing as this still relies on existing auto-model classes.

Should I include inside the docs to use an `AutoModel` class for better readability?

aarnphm commented:
@Narsil CI failed due to the recent `black` versioning locked?

Narsil commented on src/transformers/pipelines/__init__.py:
```diff
@@ -305,14 +305,14 @@
     elif values["type"] != "multimodal":
         raise ValueError(f"SUPPORTED_TASK {task} contains invalid type {values['type']}")
 
+PipelineRegistry = _PipelineRegistry.from_dict(supported_tasks=SUPPORTED_TASKS, task_aliases=TASK_ALIASES)
```
NIT: As a global and not a class name I would suggest using `PIPELINE_REGISTRY` instead.
Having a registry as a global variable is perfectly fine to me.

Also `from_dict` could be just a regular `__init__` maybe a bit more pythonic ?

aarnphm replied:
I don't have a big preference on this. I will update accordingly.

aarnphm replied:
Done.

Narsil commented:
Super nice PR thanks for this addition !

Left a few NITs about the structure. Feel free to ignore if you don't agree with them

sgugger commented:
Yes, you'll need to rebase on main to fix the tests. Failures are due to new releases of black and PyTorch.

aarnphm committed:
chore: PR updates

- revert src/transformers/dependency_versions_table.py from upstream/main
- updates pipeline registry to use global variables

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
> Yes, you'll need to rebase on main to fix the tests. Failures are due to new releases of black and PyTorch.

understood. Address accordingly

cc @LysandreJik @sgugger @Narsil when you guys have time.

Narsil reviewed (commented):
LGTM !

Letting a core maintainer approve.

sgugger reviewed (approved):
Good for me! Would like @LysandreJik to also have a look before merging.

LysandreJik commented:
That's very nice, love this approach! Should make it much much simpler to add custom pipelines.

I only have one request: please add tests :smile:

aarnphm committed:
tests: add tests for pipeline registry

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
cc @LysandreJik for tests. I'm thinking to add a test for log capture output, but I'm not too familiar with transformers logging structure.

sgugger commented:
If you want to test outputs, we have a util for this called `CaptureStd`. You can see an example of use in [this test](https://github.com/huggingface/transformers/blob/9fe2403bc52e342022ed132561655f84a6b6b7f3/tests/utils/test_cli.py#L30).

aarnphm commented:
> If you want to test outputs, we have a util for this called `CaptureStd`. You can see an example of use in [this test](https://github.com/huggingface/transformers/blob/9fe2403bc52e342022ed132561655f84a6b6b7f3/tests/utils/test_cli.py#L30).

Thanks. Will update accordingly.

aarnphm committed:
tests: add test for output warning.

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
tests are finished. lmk if any additional testing is required.

aarnphm committed:
chore: fmt and cleanup unused imports

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm commented on tests/pipelines/test_pipelines_common.py:
```diff
@@ -607,3 +607,57 @@ def add(number, extra=0):
+
+    @require_torch
+    def test_register_pipeline(self):
+        from transformers import AutoModelForSequenceClassification
+        from transformers.utils import is_torch_available
```
noted.

aarnphm committed:
fix: change imports to top of the file and address comments

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

sgugger commented:
Failure is flaky, so merging. Thanks again for your contribution!
